virtool/analyses/data.py

```diff
@@ -1,23 +1,391 @@
-            document["updated_at"] = document["created_at"]
+        document_before_formatting = document
+
+        sample = await self._db.samples.find_one(
+            {"_id": document["sample"]["id"]}, {"quality": False}
+        )
+
+        if not sample:
+            raise ResourceError()
 
         if document["ready"]:
-            return await virtool.analyses.format.format_analysis(self._app, document)
+            document = await virtool.analyses.format.format_analysis(
+                self._config, self._db, document
+            )
+
+        document = await processor(self._db, document)
+
+        if document["workflow"] == "nuvs":
+            document = await apply_transforms(
+                document,
+                [AttachNuVsBLAST(self._pg)],
+            )
+
+        for key in document_before_formatting:
+            if key not in document:
+                document.update({key: document_before_formatting[key]})
+
+        return Analysis(**document)
+
+    async def has_right(self, analysis_id: str, client, right: str) -> bool:
+        """
+        Checks if the client has the `read` or `write` rights.
+
+        :param analysis_id: the analysis ID
+        :param client: the client object
+        :param right: the right to check for
+        :return: boolean value
+        """
+        sample = await get_one_field(self._db.analyses, "sample", analysis_id)
+
+        if sample is None:
+            raise ResourceNotFoundError()
+
+        sample = await self._db.samples.find_one(
+            {"_id": sample["id"]},
+            ["user", "group", "all_read", "group_read", "group_write", "all_write"],
+        )
+
+        if not sample:
+            raise ResourceError()
+
+        read, write = get_sample_rights(sample, client)
+
+        if right == "read":
+            return read
+
+        if right == "write":
+            return write
+
+    async def delete(self, analysis_id: str, jobs_api_flag: bool):
+        """
+        Delete a single analysis by its ID.
+
+        :param analysis_id: the analysis ID
+        :param jobs_api_flag: checks if the jobs_api is handling the request
+        """
+        document = await self._db.analyses.find_one(
+            {"_id": analysis_id}, ["job", "ready", "sample"]
+        )
+
+        if not document:
+            raise ResourceNotFoundError()
+
+        sample_id = document["sample"]["id"]
+
+        if jobs_api_flag:
+            if document["ready"]:
+                raise ResourceConflictError()
+        else:
+            if not document["ready"]:
+                raise ResourceConflictError()
+
+        await self._db.analyses.delete_one({"_id": analysis_id})
+
+        path = self._config.data_path / "samples" / sample_id / "analysis" / analysis_id
+
+        try:
+            await run_in_thread(rm, path, True)
+        except FileNotFoundError:
+            pass
+
+        await recalculate_workflow_tags(self._db, sample_id)
+
+    async def upload(
```

Comment:
This method name makes it sound like it is for uploading an analysis. It is actually for uploading a file that is a child resource of the analysis.
```suggestion
    async def upload_file(
```